lib/rubocop/cop/rspec/named_subject.rb

```diff
@@ -70,6 +75,79 @@ def ignored_shared_example?(node)
           cop_config['IgnoreSharedExamples'] &&
             node.each_ancestor(:block).any?(&method(:shared_example?))
         end
+
+        def autocorrect(node)
+          if subject_definition?(node)
+            correct_subject_definition(node)
+          else
+            correct_subject_usage(node)
+          end
+        end
+
+        private
+
+        def named_subject?(node)
+          node.send_node.arguments?
+        end
+
+        def subject_definition?(node)
+          node.type == :block
+        end
+
+        def correct_subject_definition(node)
+          lambda do |corrector|
+            name = subject_name(node.ancestors.last)
+            next unless name
+
+            repacement = "subject(:#{name})"
+            corrector.replace(node.send_node.loc.selector, repacement)
+          end
+        end
+
+        def correct_subject_usage(node)
+          lambda do |corrector|
+            name = find_subject_usage_name(node)
+            next unless name
+
+            corrector.replace(node.loc.selector, name)
+          end
+        end
+
+        def find_subject_usage_name(node)
+          node.each_ancestor do |ancestor|
+            return subject_name(ancestor) if ancestor.parent.nil?
+
+            RuboCop::RSpec::ExampleGroup.new(ancestor).subjects.each do |subject|
+              if named_subject?(subject)
+                subject_name = subject.send_node.arguments.first.source
+                usage_name = subject_name.dup.tap {|s| s[0] = '' } # remove first ':' character
+                return usage_name
+              end
+            end
+          end
+
+          nil
+        end
+
+        def subject_name(node)
+          top_call = node.node_parts.first
+          top_call_first_arg = top_call.arguments.first
+          return if top_call_first_arg.nil?
+
+          klass_name_with_namespace = top_call_first_arg.source
+          klass_name = klass_name_with_namespace.split('::').last
+          underscore(klass_name).downcase
+        end
+
+        def underscore(camel_cased_word)
+          return camel_cased_word unless /[A-Z-]|::/.match?(camel_cased_word)
+          word = camel_cased_word.to_s.gsub("::", "/")
+          word.gsub!(/([A-Z\d]+)([A-Z][a-z])/, '\1_\2')
+          word.gsub!(/([a-z\d])([A-Z])/, '\1_\2')
+          word.tr!("-", "_")
+          word.downcase!
+          word
+        end
```

Comment:
This cop seems to do two things:
1. Detects unnamed `subject` declarations
2. Detects unnamed `subject` usages where the subject has a name

Do you think it would be easier to split it into two cops before implementing auto-correction? Is there something in common between `correct_subject_definition` and `correct_subject_usage`?

Reply:
It seems like a good idea.

I'm working or task where I need to fix several thousands cop offences. My strategy is meet new case and then create autocorrection for it. When I done, I will update the PR and refactor it.

Thanks for comments and valuable thougths! Hope get back soon 😀